src/TestFramework/TestFramework/Attributes/DataSource/DataRowAttribute.cs

```diff
@@ -84,6 +85,19 @@ public DataRowAttribute(params object?[]? data)
             : data.AsEnumerable();
 
         return string.Format(CultureInfo.CurrentCulture, FrameworkMessages.DataDrivenResultDisplayName, methodInfo.Name,
-            string.Join(",", displayData));
+            GetObjectString(displayData));
     }
+
+    /// <summary>
+    /// Recursively resolve collections of objects to a proper string representation.
+    /// </summary>
+    private static string GetObjectString(IEnumerable<object?> enumerable) =>
+        string.Join(
+            ",",
+            enumerable.Select(x =>
+                x == null
+                ? null
+                : x.GetType().IsArray
+                    ? $"[{GetObjectString(((IEnumerable)x).Cast<object>())}]"
+                    : x.ToString()));
```

Comment:
I guess it was the analyzers pushing you this way, but this code looks super dense. Can you please unwrap it a little bit? 

Do you think it would make sense to limit the number of recursive calls we do here by passing down a counter?

Reply:
Thanks. I have refactored it, hopefully it looks cleaner and easier to understand now

Could work with a counter, sure. Would this need to be configurable in the Attribute with a default limit so it's flexible?

Reply:
Maybe it is not needed, this is for the data row only, so the data will be provided by a constant array expression directly to the attribute, and I cannot see a way to make that array recursive. Which is the only case I am worried about.